File: JAVA/Date.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package mpd;

import java.util.Arrays;

/**
 *
 * @author mohalisad
 */
public class Date {
    public static long getLocalTime(double timezone)
    {
        long now=System.currentTimeMillis() / 1000L;
        now += (long)(timezone * 3600);
        return now;
    }
    public static String now()
    {
        Date d=new Date();
        return d + " " + clock();
    }
    public static String clock()
    {
        long now = getLocalTime(new Date().month < 7 ? 4.5 : 3.5);
        long second, minute, hour;
        second = now % 60;
        minute = (now / 60) % 60;
        hour = (now / 3600) % 24;
        return hour + ":" + minute + ":" + second;
    }

    private int year;
    private int month;
    private int day;
    private int dayOfWeek;
    @Override
    protected  Date clone()
    {
        Date d = new Date();
        d.dayOfWeek = dayOfWeek;
        d.day = day;
        d.month = month;
        d.year = year;
        return d;
    }
    private void calcDayOfWeek()
    {
        int monthDays = (month - 1) * 31 - (month < 8 ? 0 : month - 7);
        dayOfWeek = (year + year / 4 + day + monthDays) % 7;
    }
    private void setToNow(double timezone)
    {
        long now = getLocalTime(timezone);
        int dif = 0;
        now = now / 3600;
        now = now / 24;
        dif = (int)(now % (365 * 4 + 1));
        year = (int)(now / (365 * 4 + 1)) * 4 + 1348;
        day = 11;
        month = 10;
        for (int i = 0; i < dif; i++)
        {
            goNext();
        }
        dayOfWeek = (int)((5 + now) % 7);
    }
    private boolean isItKabise()
    {
        int[] arr = { 1, 5, 9, 13, 17, 22, 26, 30 };
        return Arrays.asList(arr).contains(year%33);
    }
    public Date()
    {
        setToNow(3.5);
        if (month < 7) setToNow(4.5);
    }
    public Date(int y, int m, int d)
    {
        year = y;
        month = m;
        day = d;
        calcDayOfWeek();
    }
    public Date(String input) throws Exception
    {
        String[] tokens = input.split("/");
        if (tokens.length != 3) throw new Exception("Date String is invalid");
        year = Integer.parseInt(tokens[0]);
        month = Integer.parseInt(tokens[1]);
        day = Integer.parseInt(tokens[2]);
        calcDayOfWeek();
    }
    public void goNext()
    {
        day++;
        if (month < 7)
        {
            if (day > 31)
            {
                day = 1;
                month++;
            }
        }
        else
        {
            if (month != 12 || isItKabise())
            {
                if (day > 30)
                {
                    day = 1;
                    month++;
                }
            }
            else
            {
                if (day > 29)
                {
                    day = 1;
                    month++;
                }
            }
        }
        if (month > 12)
        {
            month = 1;
            year++;
        }
        dayOfWeek = (dayOfWeek + 1) % 7;
    }
    public void goPrev()
    {
        day--;
        if (day == 0)
        {
            day = 31;
            month--;
            if (month == 0)
            {
                month = 12;
                year--;
            }
            if (month > 6)
            {
                day--;
            }
            if (month == 12 && !isItKabise())
            {
                day--;
            }
            dayOfWeek = (dayOfWeek - 1) % 7;
        }
    }
    public Date prev()
    {
        Date d = clone();
        d.goPrev();
        return d;
    }
    public Date next()
    {
        Date d = clone();
        d.goNext();
        return d;
    }
    @Override
    public String toString()
    {
        return year + "/" + month + "/" + day;
    }
    public String getDayOfWeek() throws Exception
    {
        switch (dayOfWeek)
        {
            case 0:
                return "شنبه";
            case 1:
                return "یک شنبه";
            case 2:
                return "دوشنبه";
            case 3:
                return "سه شنبه";
            case 4:
                return "چهارشنبه";
            case 5:
                return "پنج شنبه";
            case 6:
                return "جمعه";
        }
        throw new Exception("DayOfWeek is wrong");
    }
}

File: PYTHON/Date.py
import time

class Date(object):
    def __getLocalTime(timezone):
        now = int(time.time())
        now += int(timezone * 3600)
        return now
    def now():
        d = Date()
        return d.toString() + " " + Date.clock()
    def clock():
        now = Date.__getLocalTime(4.5 if Date().__month < 7 else 3.5)
        second = now % 60
        minute = int(now / 60) % 60
        hour = int(now / 3600) % 24
        return str(hour) + ":" + str(minute) + ":" + str(second)
    def __init__(self,y=-1,m=-1,d=-1):
        if d==-1:
            if y==-1:
                self.__setToNow(3.5)
                if self.__month < 7:
                    self.__setToNow(4.5)
            else:
                tokens = y.split('/')
                if len(tokens) != 3:
                    raise Exception("Date string is invalid")
                self.__year = int(tokens[0])
                self.__month = int(tokens[1])
                self.__day = int(tokens[2])
                self.__calcDayOfWeek()
        else:
            self.__year = y
            self.__month = m
            self.__day = d
            self.__calcDayOfWeek()
    def __clone(self):
        d = Date()
        d.__dayOfWeek = self.__dayOfWeek
        d.__day = self.__day
        d.__month = self.__month
        d.__year = self.__year
        return d
    def __calcDayOfWeek(self):
        monthDays = (self.__month - 1) * 31 - (0 if self.__month < 8 else self.__month - 7)
        self.__dayOfWeek = (self.__year + int(self.__year / 4) + self.__day + monthDays) % 7
    def __setToNow(self,timezone):
        now = Date.__getLocalTime(timezone)
        now = now / 3600
        now = now / 24
        dif = int(now % (365 * 4 + 1))
        self.__year = int(now / (365 * 4 + 1)) * 4 + 1348
        self.__day = 11
        self.__month = 10
        self.__dayOfWeek = 0
        for i in range(0,dif):
            self.goNext()
        self.__dayOfWeek = (int)((5 + now) % 7)
    def __isItKabise(self):
        arr = [1, 5, 9, 13, 17, 22, 26, 30]
        return self.__year % 33 in arr
    def goPrev(self):
        self.__day -= 1
        if self.__day == 0:
            self.__day = 31
            self.__month -= 1
            if self.__month == 0:
                self.__month = 12
                self.__year -= 1
            if self.__month > 6:
                self.__day -= 1
            if self.__month == 12 and not self.__isItKabise():
                self.__day -= 1
            self.__dayOfWeek = (self.__dayOfWeek - 1) % 7
    def goNext(self):
        self.__day += 1
        if self.__month < 7:
            if self.__day > 31:
                self.__day = 1
                self.__month +=1
        else:
            if self.__month != 12 or self.__isItKabise():
                if self.__day > 30:
                    self.__day = 1
                    self.__month += 1
            else:
                if self.__day > 29:
                    self.__day = 1
                    self.__month += 1
        if self.__month > 12:
            self.__month = 1
            self.__year += 1
        self.__dayOfWeek = (self.__dayOfWeek + 1) % 7
    def prev(self):
        d = self.__clone()
        d.goPrev()
        return d
    def next(self):
        d = self.__clone()
        d.goNext()
        return d
    def toString(self):
        return str(self.__year) + "/" + str(self.__month) + "/" + str(self.__day)
    def getDayOfWeek(self):
        return ["شنبه","یک شنبه","دوشنبه","سه شنبه","چهارشنبه","پنج شنبه","جمعه"][self.__dayOfWeek]
